src/dso_api/dynamic_api/temporal.py

```diff
@@ -0,0 +1,144 @@
+"""Extra functionality to handle temporality.
+
+This includes things like:
+- Only retrieve certain historical versions of an object.
+- Reduce querysets/get_object logic
+"""
+from __future__ import annotations
+
+from datetime import date, datetime
+from typing import Literal, Optional, Union
+
+from django.db import models
+from django.db.models import Q
+from django.utils.timezone import make_aware, now
+from more_itertools import first
+from rest_framework.exceptions import ValidationError
+from schematools.types import DatasetTableSchema, TemporalDimensionFields
+from schematools.utils import to_snake_case
+
+
+class TemporalTableQuery:
+    """The temporal query from the request, mapped to the current table."""
+
+    table_schema: DatasetTableSchema
+    is_versioned: bool
+
+    #: Which historical version is requested.
+    version_field: Optional[str]
+    version_value: Optional[str]
+
+    #: Which date is requested
+    slice_dimension: Optional[str]
+    slice_value: Optional[Union[Literal["*"], date, datetime]]
+    slice_range_fields: Optional[TemporalDimensionFields]
+
+    def __bool__(self):
+        return self.is_versioned
+
+    def __init__(self, request, table_schema: DatasetTableSchema):
+        temporal = table_schema.temporal
+
+        self.table_schema = table_schema
+        self.is_versioned = temporal is not None
+
+        # Parsed arguments are None unless there is a match with the request
+        self.version_field = None
+        self.version_value = None
+
+        self.slice_dimension = None
+        self.slice_value = None
+        self.slice_range_fields = None
+
+        if temporal is not None:
+            # See if a filter is made on a specific version
+            self.version_field = temporal.identifier  # e.g. "volgnummer"
+            if version := request.GET.get(self.version_field):
+                self.version_value = version
+
+            # See if a filter is done on a particular point in time (e.g. ?geldigOp=yyyy-mm-dd)
+            for dimension, fields in temporal.dimensions.items():
+                if date_value := request.GET.get(dimension):
+                    self.slice_dimension = dimension  # e.g. geldigOp
+                    self.slice_value = self._parse_date(dimension, date_value)
+                    self.slice_range_fields = fields
+
+    def _parse_date(self, dimension: str, value: str) -> Union[Literal["*"], date, datetime]:
+        """Parse and validate the received date"""
+        if value == "*":
+            return "*"
+
+        try:
+            if "T" in value or " " in value:
+                return make_aware(datetime.fromisoformat(value))
```

Comment:
if there is a "T" or an " " in the ISO formatted string that `datetime.fromisoformat` already returns an aware datetime object. So why the `make_aware`?

Reply:
It doesn't look like it does:

```
>>> from django.utils.timezone import is_aware
>>> from datetime import datetime
>>> is_aware(datetime.fromisoformat("2021-11-09 12:15:00"))
False
```

Reply:
Ah, you are right. I thought `isoformat` always included the timezone, hence `fromisoformat` would be timezone aware. But that's not the case. Learned something new. :+1: